wgengine/netstack/netstack.go

```diff
@@ -405,28 +405,36 @@ func (ns *Impl) DialContextUDP(ctx context.Context, addr string) (*gonet.UDPConn
 }
 
 func (ns *Impl) injectOutbound() {
-	for {
-		packetInfo, ok := ns.linkEP.ReadContext(context.Background())
-		if !ok {
+	// Deadline is optional - thoughts?
+	d, n := time.Second*30, time.Now()
+	r, w := net.Pipe()
+	r.SetDeadline(n.Add(d))
+	w.SetDeadline(n.Add(d))
+	defer r.Close()
+	defer w.Close()
+	for packetInfo, ok := ns.linkEP.ReadContext(context.Background()); ; packetInfo, ok = ns.linkEP.ReadContext(context.Background()) {
+		if !ok { // thoughts on getting rid of this branch?
 			ns.logf("[v2] ReadContext-for-write = ok=false")
 			continue
 		}
 		pkt := packetInfo.Pkt
-		hdrNetwork := pkt.NetworkHeader()
-		hdrTransport := pkt.TransportHeader()
-
 		full := make([]byte, 0, pkt.Size())
```

Comment:
This alloc per packet is still here, so how does this help the optimization bug?

Reply:
> This alloc per packet is still here,

Unless a buffer is provided to this function to handle each packet than it will need to allocate one, no?

> so how does this help the optimization bug?

So both implementations allocate a buffer of size `full`.

```go
full := make([]byte, 0, pkt.Size())
```

The current implementation uses append - these calls will re-allocate `full` multiple times (to my understanding), in general I've thought it best to use `io.Copy`, `copy` or `subtle.ConstantTimeCopy` when populating buffers over append, but, I did not spend a decade developing the language so please do tell me if I'm wrong.

```go
full = append(full, hdrNetwork.View()...)
full = append(full, hdrTransport.View()...)
full = append(full, pkt.Data().AsRange().AsView()...)
```

My implementation we're trading this reallocation of the `full` buffer for multiple writes to the in-memory pipe.

```go
go func() {
	w.Write(pkt.Data().AsRange().AsView())
	w.Write(pkt.NetworkHeader().View())
	w.Write(pkt.TransportHeader().View())
}()
```

Then we populate the empty, already allocated buffer `full` with a single read from the Pipe, avoiding the resizing or cost associated with the use of append. 

```go
r.Read(full)
```

Like I mentioned though I didn't benchmark this code, I simply was scoping out your repository, saw the open issue and was like oh huh _maybe you could use a pipe_. I raised the PR as the tests all passed on my end and I figured you'd have some easy CI/CD for bench-marking changes to `netstack`.

Reply:
Is packet-size a constant here?
```go
pkt := packetInfo.Pkt
full := make([]byte, 0, pkt.Size())
```
Is then sum of lengths across these three values always == `pkt.Size()` as well?
```go
w.Write(pkt.Data().AsRange().AsView())
w.Write(pkt.NetworkHeader().View())
w.Write(pkt.TransportHeader().View())
```